### ISSUE TITLE
Accessibility Audit of Angular.io - Homepage only

### ISSUE THREAD

**Issue by marcysutton** [angular/angular#16005]
I was asked by the Angular team to do an accessibility audit of the new [Angular.io](https://aio-staging.firebaseapp.com/). Here is the initial **bug report:**

## **Color Contrast**
For font-sizes under 18px, the [contrast ratio](https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html) must be 4.5:1 to be visible. For large/bold text, it can be 3:1. The text and links in the footer would benefit from a darker background, or bold text / larger font size so they are more legible to people with low vision. And yes, there are low-vision web developers!

### 21 Color Contrast Violations
- DEVIntersection Europe Register Button: `Element has insufficient color contrast of 1.38 (foreground color: #ffffff, background color: #dbdbdb, font size: 10.5pt, font weight: normal)`
- 14 Footer Links: `Element has insufficient color contrast of 4.41 (foreground color: #fafafa, background color: #1976d2, font size: 10.5pt, font weight: normal)`
- 6 White Footer Headings/Paragraphs - I audited these manually, since the background image seems to interfere. The lightest part of the background image causes a ratio just under what's accepted for WCAG 2.0 AA: http://leaverou.github.io/contrast-ratio/#hsla%28200%2C0%25%2C0%25%2C.7%29-on-%23237cd4

## Misc HTML Violations

- [x] The <html> element does not have a lang attribute
- [x] Form element missing a label: `<input placeholder="Search">` - ideally this would have a visible, paired label to give it a larger hit area, but `aria-label` would work too

## Images Missing Alt Text (6)

- [x] `<img src="assets/images/logos/angular/angular.svg" class="hero-logo">` - Should say `alt="Angular"`
- [x] `<img src="assets/images/home/responsive-framework.svg">` - Can be `alt="responsive framework"`. Has a hover style, so it should have an alt.
- [x] `<img src="assets/images/home/speed-performance.svg">` - Can be `alt="Speed and performance"` since it has a hover style
- [x] `<img src="assets/images/home/joyful-development.png">` - Can be `alt="IDE example"`
- [x] `<img src="assets/images/home/loved-by-millions.png">` - Can be `alt="Angular on the map"`
- [x] `<img src="assets/images/logos/anglebrackets/devintersection.png" width="64">` - Can be empty: `alt=""`

## Manual Audit - General

- [x] Headings - After the h1, the headings NOT in Search Results should be h2's, given the information hierarchy.
- [x] `DIV.mat-sidenav-content` should have `role="main"` to indicate the main content area as a landmark for screen reader users.
- [x] DEVIntersection button - needs a focus style in addition to more contrast.

## Sidenav
- [x] Sidenav - Needs focus management on open and close. Keyboard users can't get to the menu easily.
- [x] Sidenav Links - No href attributes, so they aren't focusable. I'd recommend buttons for toggles instead of links.

## Search Results

- [x] Give this region a visually hidden heading, something like "h2. Search Results". The other headings can become h3's.
- [x] Use unordered lists so the number of links will be counted
- [x] There is currently no easy way for a keyboard user to get to the search results, since they come after the rest of the DOM content. Perhaps it should be moved to come after the search field and use CSS positioning to make it a sibling in the DOM? 
- [x] Escape key should send focus back to the triggering input

Let me know if you have any questions.

**Comment by sjtrimble** [comment]
@marcysutton Regarding "DEVIntersection button - needs a focus style in addition to more contrast." do you mean the "Register Now" button that has a focus style or something else?

**Comment by marcysutton** [comment]
@sjtrimble oh yes! That's what I meant. The "Register Now" button in that callout region.

**Comment by sjtrimble** [comment]
@marcysutton Your comments asks for the addition of a focus style, but it already has this. I want to make sure I'm not missing anything here. Can you confirm? Thanks a bunch!

**Comment by marcysutton** [comment]
@sjtrimble are you looking at a different build perhaps? The site linked above in my ticket definitely does not have a focus style for that link button–it has hover only. Here's a screenshot:

![Register Now link button open in developer tools](https://cloud.githubusercontent.com/assets/1045233/25205693/ccbf6772-2518-11e7-9f21-e5dff8691e9f.png)

**Comment by sjtrimble** [comment]
@marcysutton My error!! Thanks for clarifying. Will look into that :)

**Comment by wardbell** [comment]
@marcysutton regarding
>Give this region a visually hidden heading, something like “h2. Search Results”. The other headings can become h3's.

I see that @sjtrimble did add the `visually-hidden` class in `search-results.component.html`. I still don’t understand how this works, what it solves, and whether we should apply similar reasoning elsewhere. Can you shed some light on this?

**Comment by wardbell** [comment]
"DevIntersection Button" is checked "done" because it no longer exists.

**Comment by marcysutton** [comment]
@wardbell the idea is to provide a heading to the Search Results section for screen reader users, because they often navigate by headings. It could just as easily be a regular visible heading, but a heading with "visually-hidden" CSS preserves the existing design by hiding it from sighted users yet providing semantic structure for screen reader navigation. Making this new heading an h2 and restructuring the other Search Result headings as h3's will make it into a proper content hierarchy.

There are multiple ways to code visually hidden CSS; [this method](https://github.com/angular/angular/blob/master/aio/src/styles/1-layouts/_layout-global.scss#L43) should work fine.

**Comment by wardbell** [comment]
@marcysutton Can we use `<section>`/`<article>` instead of `<h2>/<h3>`? It's not something that you can link to. It's just a label for a collection of results (if there happen to be any).

**Comment by wardbell** [comment]
PR #16468 Moves search results under search box (just below the top toolbar).  The first tab out of the search box ends up in the first result. I believe that's what was necessary to check a box in the Search Results section above.

**Comment by wardbell** [comment]
@marcysutton regarding Search Results item:
>Escape key should send focus back to the triggering input
I think it does that now. When I type in the search box and hit escape, I stay in the search box. 

What behavior are we missing? If I mouse or tab out of the search box and then hit escape, I do _not_ return to the search box. Should I? Is that what you mean?

**Comment by gkalpak** [comment]
I am pretty sure that's what she meant :smiley:

**Comment by petebacondarwin** [comment]
Fixed

> Escape key should send focus back to the triggering input

in #16758

**Comment by petebacondarwin** [comment]
Fixed:

>  Sidenav Links - No href attributes, so they aren't focusable. I'd recommend buttons for toggles instead of links.

in https://github.com/angular/angular/pull/16807

**Comment by petebacondarwin** [comment]
All the checkboxes are ticked!

**Comment by marcysutton** [comment]
Noting that there may still be additional issues on other pages in the absence of a full audit, I found another serious item. The `header-link` anchor tags used to deep link headings are poorly constructed:

The anchors have `aria-hidden="true"` but they still have hrefs, causing what we call "ghost controls" with no accessibility information when focused. Pair that with a hover state on the child icon element only and no focus state to make them visible, and they are pretty bad accessibility-wise.

Here's a screenshot:
![aio-header-links](https://cloud.githubusercontent.com/assets/1045233/26172761/5c128526-3afe-11e7-9f0f-def052b36ab5.png)

To fix these, I'd move the hover state to the anchor and add a focus state at the same time. Remove `aria-hidden="true"` and they should work better.

**Comment by petebacondarwin** [comment]
Moving the issue with the heading anchors to a new issue: https://github.com/angular/angular/issues/16854
